lib/urlwatch/reporters.py

```diff
@@ -668,18 +668,88 @@ def submit(self):
             summary.extend(summary_part)
             details.extend(details_part)
 
+        # Sum of the lengths of parts, with newlines taken into account
+        summary_len = sum(len(part) for part in summary) + len(summary) - 1
+
+        # Determine the space remaining after the summary. Reserve some extra
+        # space so we can at least report that all details were omitted if the
+        # message is too long.
+        if max_length is not None:
+            remaining_len = max_length - summary_len - 40
```

Comment:
Could you combine all the `max_length is not None` blocks into a single block below?

Or maybe alternatively move the max length code into its own function that takes all the details part and does the length distribution?

Reply:
In other words, can you make it so that the code is easy to read/understand if `max_length is None` (right now one has to read lots of code and all the `max_length` conditions are not hit).

Reply:
This block is here on its own because we're using `remaining_len` in the block below when we calculate the footer length so we need it set by then.

I don't like the current state of the code either, I'll see what I can do. Some tests might be nice too since this is tricky.

Reply:
I refactored the code into a separate `_render` method which renders the report parts, trimming them down if necessary and reporting whether anything was trimmed. The code is now much easier to follow (IMO).